### ISSUE TITLE
Content Resolution And Gateway Performance

### ISSUE THREAD

**Issue by Stebalien** [ipfs/go-ipfs#6383]
Finding content on the IPFS network is often quite slow. This meta-issue tracks the issue, the reasons behind it, and the work being done to fix it.

There are two main causes for not being able to find content on the IPFS network:

1. There are either no IPFS peers providing the content or those peers are unreachable (e.g., they're behind a NAT).
2. The DHT is slow and unreliable.

The first issue is being addressed with better NAT traversal:

* AutoRelay (https://github.com/ipfs/go-ipfs/issues/6290): Automatically use a relay if the IPFS node is behind a NAT.
* Hole Punching ([TCP](https://en.wikipedia.org/wiki/TCP_hole_punching), [UDP](https://en.wikipedia.org/wiki/UDP_hole_punching)). (no tracking issue yet)

The second issue is more complicated:

1. The primary issue is that most of the DHT is unreachable (behind NATs). That means every DHT query spends a significant amount of time trying to contact unreachable peers. This situation can be improved by preventing NATed nodes from joining the DHT (https://github.com/libp2p/go-libp2p-kad-dht/issues/216, https://github.com/libp2p/go-libp2p-kad-dht/issues/330).
2. Another issue is that the DHT is unreliable. That is, many of the nodes in the DHT are ephemeral so the DHT quickly forgets the content it's hosting. Fixing https://github.com/libp2p/go-libp2p-kad-dht/issues/216 would help solve this issue as well.
2. The current DHT implementation doesn't correctly implement the Kademlia protocol: https://github.com/libp2p/go-libp2p-kad-dht/pull/291. This makes queries take even _longer_ than they should.

**Comment by Stebalien** [comment]
cc @vyzo, @raulk am I missing anything here?

**Comment by jasonzhouu** [comment]
Maybe the suggestion in https://discuss.ipfs.io/t/proposal-peer-hint-uri-scheme/4649 is feasible.
> Instead, a more rational way for this would be to optionally add to the URI a hint where the file is available, i.e. which peer is the originator. This would hasten the download while the file isn’t popular yet.

**Comment by anshbansal** [comment]
> Maybe the suggestion in https://discuss.ipfs.io/t/proposal-peer-hint-uri-scheme/4649 should be considered, i.e. giving users an option to specify the location of file if the file is not popular yet, so as to make DHT resolution much faster.
> 
> > Instead, a more rational way for this would be to optionally add to the URI a hint where the file is available, i.e. which peer is the originator. This would hasten the download while the file isn’t popular yet.
> 
> It can solve the issue of #6382 , as the user @voxsoftware knows where the file located exactly.

There is this pinning service called Pinata https://pinata.cloud/documentation#AddHashToPinQueue which allows me to specify upto 5 addresses where my content is already available. Maybe they have a partial solution to this problem already?

**Comment by Stebalien** [comment]
That just punts the problem. IPFS is supposed to be a decentralized, content-addressed network. I agree we should consider adding, e.g., a header to gateway requests that says "please connect to this specific peer when trying to find this content" but that doesn't really fix the underlying issue.

**Comment by jasonzhouu** [comment]
>There is this pinning service called Pinata https://pinata.cloud/documentation#AddHashToPinQueue which allows me to specify upto 5 addresses where my content is already available. Maybe they have a partial solution to this problem already?

@anshbansal Thank you for the reply. Yeah, specifying "multiaddresses" where the file already exists when pinning file in pinata is very useful. But it is used for the purpose that **help Pinata find your content faster**, which is not that general enough. It will be better if this feature can be added to IPFS.

**Comment by jasonzhouu** [comment]
> ipfs add a file and then fetch it from another node.
@Stebalien Why add a file, and also fetch it?

**Comment by Stebalien** [comment]
@jasonzhouu fixed that. My point is that the suggestion doesn't fix the case where an ordinary user adds a file to their IPFS node and then tries to view it on the gateway.

**Comment by jasonzhouu** [comment]
> the suggestion doesn't fix the case where an ordinary user adds a file to their IPFS node and then tries to view it on the gateway.

Yeah, it's true. If we can specify multiaddress for the node that already has the content, the gateway can fetch it faster, like what pinata does.

**Comment by kivutar** [comment]
I can confirm this issue. Testing with 3 nodes:
- an EC2 VM based in US (port 4001 open)
- a dedicated server based in Europe (port 4001 open)
- Pinata (they could pin my hash and I can resolve the hash on their gateway)

I'm unable to access my hash on ipfs.io or ipfs.infura.io hours after.

**Comment by anshbansal** [comment]
> That just punts the problem. IPFS is supposed to be a decentralized, content-addressed network. I agree we should consider adding, e.g., a header to gateway requests that says "please connect to this specific peer when trying to find this content" but that doesn't really fix the underlying issue.
>
> Basically, we won't consider this fixed until a user can `ipfs add` a file and then fetch it from ~another node~ the gateway.

That would help make things faster. A slight variation could be to add an option to ipfs companion extension that allows us to specify the value of the header for the requests sent to the ipfs gateway.  Or perhaps an API in the local IPFS daemon that the browser extension can use to see if the hash is present and add my address to the header automatically instead of the routing option currently present? This would also be helpful for the gateways if this is standardised as the searching for content would be easier in that case which might make operating gateways cheaper due to the reduced bandwidth costs. 

"we won't consider this fixed" sounds great. Ultimately having it at the protocol level is the perfect solution. 

> @jasonzhouu fixed that. My point is that the suggestion doesn't fix the case where an ordinary user adds a file to their IPFS node and then tries to view it on the gateway.

It does fix it to some extent. If the header automatically includes my address and disable the routing to local node then it should make it faster as the gateway knows where the hash is present.

**Comment by godcong** [comment]
@Stebalien 
About Hole Punching (TCP, UDP). (no tracking issue yet)

I tested on my network, connect to the source ip+port will return error: connect: connection refused.
but, i tested with reuse.Dial + NAT with go `net.Conn` package.
connection and data transfer are normal.

**Comment by DonaldTsang** [comment]
I am thinking of some kind of "reputation/trus algorithm" relevant to https://github.com/ipfs/go-ipfs/issues/6097 ?
https://arxiv.org/ftp/arxiv/papers/1411/1411.3294.pdf

**Comment by aarshkshah1992** [comment]
@Stebalien What's the "status" of  libp2p/go-libp2p#188 ? Is there anything blocking us from going ahead with it now that we pion has full fledged WebRTC support ?

**Comment by aarshkshah1992** [comment]
@jacobheun Assigning this to myself as this mostly deals with NATs/connectivity.

**Comment by Stebalien** [comment]
I don't know bug @jacobheun probably does.

**Comment by dokterbob** [comment]
As I've stated in #5541, at ipfs-search.com we've been structurally seeing around 40-60% timeout rates for *fresh hashes*. Part of this is due to us not (yet) using streaming listing - but it seems big part does seem to relate to go-ipfs (still) not being able to find content (although, for a while, after switching to 0.5.0 we suddenly had a much better rate - I'm still trying to figure out why that was - but we need time to gather reliable data).

**Comment by jacobheun** [comment]
> What's the "status" of libp2p/go-libp2p#188 ? Is there anything blocking us from going ahead with it now that we pion has full fledged WebRTC support ?

Circling back on this we'll be starting with QUIC hole punching work and then expanding from there. There are several things we need to do in libp2p to make this work well. Once we have other things in place like direct connection upgrading and quic hole punching, we'll start looking at webrtc to expand our ability to connect to nodes not using quic.

> although, for a while, after switching to 0.5.0 we suddenly had a much better rate

**Finding** content query times and success rates on the network were drastically improved in 0.5 and later, being able to **retrieve** that content is the next step we're working on solving which should create another large boost to getting content.

**Comment by eminence** [comment]
Some additional debugging tools would be useful I think.  I have two nodes running 0.7.0, and both are directly connectable (i.e. they are either not behind a NAT, or they are behind a NAT with port forwarding).    Running `ipfs ls` on one node (to get content available on the other node) has been stuck for hours.  If I manually connect the nodes (with `ipfs swarm connect`), the content is fetch within seconds.  This makes me think something funny is going on in the DHT, but I don't really have the tools to debug it.

**Comment by dokterbob** [comment]
@eminence Direct peering should provide a workaround for such issues https://github.com/ipfs/go-ipfs/blob/master/docs/config.md#peering

**Comment by SomajitDey** [comment]
I experienced significant increase in content resolution and gateway performance using the tcp reverse proxy provided by ngrok. After exposing my laptop's tcp://localhost:4001 through ngrok, the two gateways: https://ipfs.io/ipfs and https://gateway.pinata.cloud/ipfs/ found my locally hosted file within minutes after doing `ipfs add`.

The method is detailed at https://gist.github.com/SomajitDey/25f2f7f2aae8ef722f77a7e9ea40cc7c

**Comment by lidel** [comment]
Closing as mostly addressed:

- DHT performance: Resolved in v0.5-v0.8+ + v0.38 will have #10834 
- Gateway timeout UX: Significantly improved in v0.37/38 with configurable timeouts, enhanced error pages with diagnostics, and retry functionality: #10950 
- Provider performance: Will be improved by #10834 (opt-in for v0.38, targeting v0.39 as default) 

The original meta-issue has served its purpose. For specific remaining concerns (NAT traversal, connectivity), please open targeted issues as Kubo 0.38+ continues to evolve.
